supervisor/process.go

```
@@ -0,0 +1,408 @@
+	if err := p.cmd.Start(); err != nil {
+		// process failed to start.
+		// exit immediately
+		p.status.StartTime = now.UnixNano()
+		p.status.StopTime = time.Now().UnixNano()
+		p.status.Error = err
+		p.setState(Fatal)
+		p.statusChan <- p.status
+		close(p.done)
+
+		return p.statusChan
+	}
+
+	p.status.StartTime = now.UnixNano()
+	p.status.PID = p.cmd.Process.Pid
+	p.setState(Running)
+	go p.wait()
+
+	return p.statusChan
+}
+
+// wait waits for a process to finish running and sets it's final status
+func (p *Process) wait() {
+	// send the final status and close done channel to signal to
+	// goroutines the process has finished running
+	defer func() {
+		p.statusChan <- p.GetStatus()
+		close(p.done)
+	}()
+
+	err := p.cmd.Wait()
+	now := time.Now()
+	exitCode := 0
+
+	p.Lock()
+	defer p.Unlock()
+
+	// get the exit code of the process
+	if err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			err = nil
+
+			if status, ok := exitErr.Sys().(syscall.WaitStatus); ok {
+				exitCode = status.ExitStatus()
+
+				if status.Signaled() {
+					err = errors.New(exitErr.Error())
+					p.setState(Interrupted)
+				}
+
+			}
+		}
+	}
+
+	// set final status of the process
+	p.status.Error = err
+	p.status.ExitCode = exitCode
+	p.status.StopTime = now.UnixNano()
+	p.setState(Finished)
+}
+
+// GetStatus returns the current status of the process
+func (p *Process) GetStatus() Status {
+	p.Lock()
+	defer p.Unlock()
+
+	return p.status
+}
+
+// Stop stops a process by sending it's group a SIGTERM signal
+func (p *Process) Stop(byUser bool) error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.IsInitialState() || p.IsFinalState() {
+		return nil
+	}
+
+	p.setState(Stopping)
+	p.status.StopByUser = byUser
+
+	return syscall.Kill(-p.cmd.Process.Pid, syscall.SIGTERM)
+}
+
+// Signal sends an os signal to the process
+func (p *Process) Signal(sig syscall.Signal) error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.IsInitialState() || p.IsFinalState() {
+		return nil
```

This should probably return an error like "Failed to signal process, already in Initial/Final state"